src/zeit/content/volume/browser/toc.py

```diff
@@ -0,0 +1,392 @@
+
+    def _get_metadata_from_article_xml(self, atricle_tree):
+        """
+        Get all relevant normalized metadata from article xml tree.
+        :param atricle_tree: lxml.etre Element -
+        :return: {'page': str, 'author': str, 'title': str, 'teaser': str}
+        """
+        xpaths = {
+            'title': "body/title/text()",
+            'page': "//attribute[@name='page']/text()",
+            'teaser': "body/subtitle/text()",
+            'author': "//attribute[@name='author']/text()"
+        }
+        res = {}
+        for key, xpath in xpaths.iteritems():
+            res[key] = atricle_tree.xpath(xpath)
+        return self._normalize_toc_element(res)
+
+    def list_relevant_ressort_dirs_with_dav(self, path):
+        """ Returns all relevant ressort paths to directories for a given path """
+        try:
+            response = self.client.propfind(path, depth=1)
+            assert response.is_multistatus
+            return [element.href for element in response if self._is_relevant_path_to_directory(path, element)]
+        except HTTPUserError:
+            return []
+
+    def _is_relevant_path_to_directory(self, root_path_of_element, element):
+        """
+
+        :param root_path_of_element: root path of DAV archive
+        :param element: tinydav status element
+        :return: bool
+        """
+        try:
+            # TODO Put this in the excluder ?
+            folders_to_exclude = {'images', 'leserbriefe'}
+            folders_to_exclude = set.union(folders_to_exclude, {ele.title() for ele in folders_to_exclude})
+            root_paths = {root_path_of_element, '/' + root_path_of_element}
+            return self._is_dav_dir(element) \
+                   and not any(folder in element.href for folder in folders_to_exclude) \
+                   and element.href not in root_paths
+        except:
+            raise
+
+    def _get_all_files_in_folder(self, ressort_path):
+        response = self.client.propfind(ressort_path, depth=1)
+        assert response.is_multistatus
+        return [status_element.href for status_element in response if not self._is_dav_dir(status_element)]
+
+    def _is_dav_dir(self, status_element):
+        try:
+            return 'directory' in status_element.get('getcontenttype').text
+        except AttributeError:
+            return False
+
+    def _create_csv(self, toc_data):
+        """
+        Creates CSV File from TOC Data.
+        SEITENZAHL [tab] AUTOREN [tab] TITEL + TEASER
+        :param toc_data: The Toc data as ordered dict.
+        :return: unicode - csv content
+        """
+
+        file_content = u''
+        out = StringIO.StringIO()
+        try:
+            writer = csv.writer(out, delimiter=self.CSV_DELIMITER)
+            for toc_element in self._generate_csv_rows(toc_data):
+
+                writer.writerow(
+                    [val.encode('utf-8') for val in toc_element]
+                )
+
+            file_content = out.getvalue()
```

Comment:
Hier hatte ich leider auch Probleme mit zope.file.file.File. So wie ich das verstanden habe muss, ich in der ZODB das File zwischenlagern (und dann transaction.commit()) bin mir aber nicht sicher wo dass dann hingelegt werden kann. Benutzte ich das Repository und lege das dort hin
brauche ich eine unique-ID. Ich würde vorschlagen, dass einfach so zu belassen, da das hacky wird (unique-ID ans zope-Fileobject) wenn ich das mache, außer du hast eine Idee oder Stelle wo ich mir das gut abgucken kann.